lightning/src/ln/channelmanager.rs

```diff
@@ -9123,17 +9152,21 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
 
 		let nonce = Nonce::from_entropy_source(entropy);
 		let context = OffersContext::InvoiceRequest { nonce };
-		let path = $self.create_blinded_paths_using_absolute_expiry(context, absolute_expiry)
-			.and_then(|paths| paths.into_iter().next().ok_or(()))
+
+		let mut builder = OfferBuilder::deriving_signing_pubkey(node_id, expanded_key, nonce, secp_ctx)
+			.chain_hash($self.chain_hash);
+
+		if let Some(path_type) = blinded_path {
+			let paths = match path_type {
+				BlindedPathType::Compact => $self.create_compact_blinded_paths(context),
+				BlindedPathType::Full => $self.create_blinded_paths(MessageContext::Offers(context)),
+			}
 			.map_err(|_| Bolt12SemanticError::MissingPaths)?;
-		let builder = OfferBuilder::deriving_signing_pubkey(node_id, expanded_key, nonce, secp_ctx)
-			.chain_hash($self.chain_hash)
-			.path(path);
 
-		let builder = match absolute_expiry {
-			None => builder,
-			Some(absolute_expiry) => builder.absolute_expiry(absolute_expiry),
```

Comment:
Could you be explicit about what you want the `MessageRouter` interface to look like? Feels like we are going back and forth without much progress...

Reply:
I think this approach makes sense!

Assuming the `scid` lookup is almost free, if `MessageRouter` has access to `MessageForwardNode` and `WhyWeWantABlindedPath`, it can make the right choice between creating a full-length or compact blinded path.

Additionally, if it’s the main decider on path compactness, it could manage the following setting:

```rust
for path in &mut paths {
	path.use_compact_introduction_node(&network_graph);
}
```

This way, it can help avoid creating a compact last hop for full-length paths.

To support the three cases Jeff mentioned in his [comment](https://github.com/lightningdevkit/rust-lightning/pull/3246#discussion_r1805066862), we could consider updating the `create_offer_builder` input fields like so:

```rust
fn create_offer_builder(..., why: Option<WhyWeWantABlindedPath>) -> Offer {
	...
}
```

1. Passing `None` supports creating an offer without a blinded path.
2. Passing `WhyWeWantABlindedPath::ForOffer { offer_expiry_time, is_size_constraint: true, ... }` covers the offer for QR codes.
3. Passing `WhyWeWantABlindedPath::ForOffer { offer_expiry_time, is_size_constraint: false, ... }` covers the unconstrained setting (e.g., an offer on wire).

@TheBlueMatt

I do have a couple of things I’m curious about with this approach:

→ Since `MessageRouter` is a `pub` trait and could be implemented outside of LDK, what if someone uses it in a setting where `scid` lookup for creating `MessageForwardNode` is resource-intensive?

→ Would it be useful for `MessageRouter` to send a note upstream on which type of Blinded Path it ultimately created?

Reply:
Seems like there would be some duplication with `MessageContext`. And if a new variant is added to `MessageContext`, then one would need to be added to `WhyWeWantABlindedPath`, too. How about:

```rust
pub enum BlindedPathUse {
    Persistent,
    Ephemeral { expiry: Duration },
    ReplyPath,
}
```

That way were aren't tying it strictly to existing BOLT12 use cases.

Reply:
Oh, but that won't be `absolute_expiry` won't be sufficient for saying no blinded paths. 😕

Reply:
> Hmm, maybe? I was thinking about the context because someone might want to have a different policy for refunds vs offers, no matter the specific expiry, but maybe we don't need to support that?

It's still possible to differentiate using the passed `MessageContext`.

> Indeed, supporting non-BOLT12 is a bit difficult, though, so maybe the more generic terminology is right.

Yeah, I think I'd prefer that, but we still need to figure out the `ChannelManager` API. We'll need to continue to pass the absolute expiry, but now we need another parameter to indicate whether to include any blinded paths. Using `Option<BlindedPathUse>` won't work since `BlindedPathUse::ReplyPath` isn't applicable.

We could drop that variant in favor of using `Persistent` for both that and long-lived offers/refund given we don't need to use the compact representation for reply paths, even though their lifetime is ephemeral.

Reply:
> I'm not sure I'm convinced it needs to change. The interface currently allows someone to decide to/not to include a blinded path based on the type of thing we're building, so the question is whether someone is going to want to generate both blinded and non-blinded versions of the same object(s) at the same time in the same node. I can see someone wanting an option to switch it, but that option can be in their route-wrapper, the question is if they will hit races in doing so.

Regarding races, seems we could make methods like `create_offer_builder_using_router` in order to pass a custom router. `DefaultMessageRouter` is cheap to construct, so I can imagine we could also provide methods on it to customize it for one-time use (i.e., by wrapping it). For example:

```rust
let router = DefaultMessageRouter::new(network_graph, entropy_source).without_blinded_paths();
let offer = channel_manager.create_offer_builder_with_router(router).build().unwrap();
```

Or provide a separate `MessageRouter` implementation that returns no paths. Either approach would have equivalent behavior.

> If a user does want this, however, they might also want some different path selection that isn't just "blinded path or not", they might want more/fewer/longer/shorter/compact/non-compact paths based on the specific call. The only way to allow for that, I think, is to pass some kind of opaque "ID" through the `ChannelManager` to the router which the user can map on.

Passing a custom router would be cleaner, IMO, given it can be essentially stateless. `ChannelManager` only needs to hold one for replies. Though arguably a future `OffersMessageFlow` would own that instead. In that world, we could have the `OffersMessageFlow`'s user-provided trait parameterization specify how to create reply routes. It would make it such that `OffersMessageFlow` wouldn't need a direct `MessageRouter` parameterization, though that may not be an entirely necessary thing to do.

Reply:
> Hmm, not quite visualizing this one but ISTM that we'd want `OffersMessageFlow` to do most of the things `ChannelManager` does today, mostly just moving the code into the new flow manager.

Suppose that `OffersMessageFlow` was parameterized by some `ResponseStrategy` trait (for lack of a better name), which the user must either implement on their own or use a `DefaultResponseStrategy` provided by LDK. The trait could be responsible not only for whether a response should be sent (e.g., if the `InvoiceRequest`'s msat-denominated amount is sufficient for a fiat-denominated offer) but also for how to construct the reply path (i.e., by returning a specific `MessageRouter` as defined by an associated type on the trait).

Reply:
FWIW, I'm not convinced we need to do what I suggested. Reply path construction doesn't really need much information outside what we'd already provide the `MessageRouter` with the `MessageContext`. I was mostly thinking out loud about how we might remove the `MessageRouter` trait entirely from `OffersMessageFlow`. For creation / initiation uses, the user would provide it directly to whichever method. For the handler, it would need to come from the parameterized `ResponseStrategy`. But that simply moves the parameterization from `OffersMessageFlow` to an associated type on `ResponseStrategy`. So probably doesn't really gain us much.

Reply:
> Can we just straight up give the `MessageRouter` the `Offer` we're replying to?

`MessageContext` is probably better as it would include any new uses.